README.md

<div align="center">
	<p>
		<img alt="Thoughtworks Logo" src="https://raw.githubusercontent.com/ThoughtWorks-DPS/static/master/thoughtworks_flamingo_wave.png?sanitize=true" width=200 />
    <br />
		<img alt="DPS Title" src="https://raw.githubusercontent.com/ThoughtWorks-DPS/static/master/dps_lab_title.png?sanitize=true" width=350/>
	</p>
  <h3>DPS Convenience Images</h3>
  <h1>twdps/di-circleci-remote-docker</h1>
  <a href="https://app.circleci.com/pipelines/github/ThoughtWorks-DPS/di-circleci-remote-docker"><img src="https://circleci.com/gh/ThoughtWorks-DPS/di-circleci-remote-docker.svg?style=shield"></a> <a href="https://hub.docker.com/repository/docker/twdps/di-circleci-remote-docker"><img src="https://img.shields.io/docker/v/twdps/di-circleci-remote-docker?sort=semver"></a> </a> <a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/github/license/ThoughtWorks-DPS/di-circleci-remote-docker"></a>
</div>
<br />

With inspiration from the CircleCI convenience images, `twdps/di-circleci-remote-docker` has both alpine and buster-slim variants created with self-hosted runners in mind. As the name suggests, this image is designed to serve as a starter image for building a use-tailored CircleCI [remote docker executor](https://circleci.com/docs/2.0/custom-images/#section=configuration).  

This image contains the minimum packages required to operate a build on CircleCI.  

_difference with cimg libraries._ Enterprise settings often require specific security and configuration testing. The twdps series of convenience images includes common sdlc security practices, including CIS-benchmark testing. The Alpine image will generally not have any cve issues.  

**Other images in this series**  

twdps/di-circleci-base-image  
twdps/di-circleci-infra-image  

## Table of Contents

- [Getting Started](#getting-started)
- [What is included in the image](#what-is-included-in-the-image)
- [Development](#development)
- [Contributing](#contributing)
- [Additional Resources](#additional-resources)

## Getting Started

This image intended to be used as the FROM image in a custom CircleCI remote docker executor.  

For example:

```Dockerfile
FROM twdps/di-circleci-remote-docker:2020.06   # alpine:3.12.0

ENV NODE_VERSION=12.16.3

RUN curl -L -o node.tar.xz "https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-x64.tar.xz" && \
	sudo tar -xJf node.tar.xz -C /usr/local --strip-components=1 && \
	rm node.tar.xz && \
	sudo ln -s /usr/local/bin/node /usr/local/bin/nodejs
```

The tag `2020.05` indicates that the version of the base image used will be the May 2020 release.  
See how tags work below for more information.  

## What is Included in the Image

This image is based on the Alpine Linux distribution and contains the minimum requirements needed to be used as a remote docker executor on CircleCI:  

- git
- openssh
- tar
- gzip
- ca-certificates
_See CHANGES.md for current versions_

_Note. Earlier versions of this executor included sudo, multi-language libraries, and defined USER=circleci. This functionality has been moved to twdps/di-circleci-base-image starting with 2021.07_  

_Note. Multiple cve results in scan of slim image. See CHANGES.md for details._

### Tagging Scheme

This image has the following tagging scheme:

```
twdps/di-circleci-remote-docker:edge
twdps/di-circleci-remote-docker:stable
twdps/di-circleci-remote-docker:<YYYY.MM>
```

`edge` - points to the latest version of the Base image. Built from the `HEAD` of the `master` branch. Intended to be used as a testing version of the image with the most recent changes however not guaranteed to be all that stable and not recommended for production software.  

`stable` - points to the latest, production ready base image. For projects that want a decent level of stability while recieving all software updates and recommended security patches. Security patches can sometimes include pre-release or release candidate versions of packages. Consider this similar to using the `:latest` tag and is not a generally recommended practice. Pin the `FROM` reference to a specific release and adopt new releases as part of your ci process. Typically updated once a month.  

`<YYYY.MM>` - Release version of the image, referred to by the 4 digit year, dot, and a 2 digit month. For example `2020.05` would be the monthly tag from May 2020. This is the recommended version for use in an executor Dockerfile.  

## Development

Images can be built and run locally with this repository.  
This has the following requirements:

- local machine of Linux (Alpine tested) or macOS
- modern version of a shell (zsh/bash tested (v4+))
- modern version of Docker Engine (v19.03+)

### Building the Dockerfiles

To build and test the Docker image locally, run the `testlocal.sh` script:

```bash
./testlocal.sh
```

*requirements for testing*  
conftest  
bats  

### Publishing Official Images (for Maintainers only)

Git push will trigger the dev-build pipeline. In addition to the tests performed in testlocal.sh, a snyk scan is done to expose any known vulnerabilities.  

To create a release version, simply tag HEAD with the release version format `YYYY.MM`  

## Contributing

We encourage [issues](https://github.com/twdps/di-circleci-remote-docker/issues) and [pull requests](https://github.com/twdps/di-circleci-remote-docker/pulls) against this repository. In order to value your time, here are some things to consider:

1. Intended to be the minimum configuration necessary for an alpine-based image to be successfully launched by circleci as a remote docker executor and specifically does not include any other packages. As such, the way to use the image is in the `FROM twdps/di-circleci-remote-docker:tag` statement of your Dockerfile.
1. PRs are welcome. Given the role of this image as a building block in building CircleCI remote docker executors, it is expected that PRs or Issue will be releated to bugs or compatibility issues. PR's to include additional packages will only be considered where necessary to continue supporting Alpine linux as a remote docker base.  

## Additional Resources

- [CircleCI Docs](https://circleci.com/docs/)  
- [CircleCI Configuration Reference](https://circleci.com/docs/2.0/configuration-reference/#section=configuration)  


CHANGES.md
## 07-02-2021 package versions in di-circleci-remote-docker 2021.07.1

^change

no changes in alpine-2021.07.1

FROM debian:bullseye-20210621-slim

| package         | version           |
|-----------------|-------------------|
| git             | 1:2.30.2-1^       |
| openssh-server  | 1:8.4p1-5^        |
| tar             | 1.34+dfsg-1^      |
| gzip            | 1.10-4^           |
| ca-certificates | 20210119^         |


## 07-02-2021 package versions in di-circleci-remote-docker 2021.07

*changes

FROM alpine:3.14.0

| package         | version     |
|-----------------|-------------|
| git             | 2.32.0-r0   |
| openssh         | 8.6_p1-r2   |
| tar             | 1.34-r0     |
| gzip            | 1.10-r1     |
| ca-certificates | 20191127-r5 |

FROM debian:buster-20210621-slim

| package         | version           |
|-----------------|-------------------|
| git             | 2.20.1-2+deb10u3* |
| openssh-server  | 7.9p1-10+deb10u2* |
| tar             | 1.30+dfsg-6*      |
| gzip            | 1.9-3*            |
| ca-certificates | 20200601~deb10u2* |


test.sh
docker run -it \
           -v /var/run/docker.sock:/var/run/docker.sock \
           -e "SNYK_TOKEN=$SNYK_TOKEN" \
           -v "${PWD}:/project" \
           snyk/snyk-cli:docker test \
           --org=twdps \
           --severity-threshold=low \
           --docker twdps/di-circleci-remote-docker:2020.05 \
           --file=Dockerfile
           


testlocal.sh
#!/usr/bin/env bash
_() { echo 'cleanup'; docker rm -f di-circleci-remote-docker-alpine-edge; docker rmi -f twdps/di-circleci-remote-docker:alpine-edge; docker rm -f di-circleci-remote-docker-slim-edge; docker rmi -f twdps/di-circleci-remote-docker:slim-edge ; }
trap _ EXIT

type=${1:-'pinned'}
alpine_dockerfile='Dockerfile.alpine'
slim_dockerfile='Dockerfile.slim'

# pass unpinned as a parameter to this script to test the unpinned
# package definitions for early detection of upcoming breaking changes
if [[ $type == "unpinned" ]]; then
  alpine_dockerfile='Dockerfile.alpine.unpinned'
  slim_dockerfile='Dockerfile.slim.unpinned'
fi

echo "build twdps/di-circleci-remote-docker:alpine"
docker build -t twdps/di-circleci-remote-docker:alpine-edge -f $alpine_dockerfile .
echo "build twdps/di-circleci-remote-docker:slim"
docker build -t twdps/di-circleci-remote-docker:slim-edge -f $slim_dockerfile .

echo "run image configuration tests"
docker run -it -d --name di-circleci-remote-docker-alpine-edge --entrypoint "/bin/ash" twdps/di-circleci-remote-docker:alpine-edge
docker run -it -d --name di-circleci-remote-docker-slim-edge --entrypoint "/bin/bash" twdps/di-circleci-remote-docker:slim-edge
bats test


profiles/circleci-remote-docker/controls/circleci-remote-docker.rb
control 'packages' do
  impact 1.0
  title 'confirm package installation'
  desc 'confirm all desired packages are installed'
  describe command('apk info -v') do
    its('stdout') { should include ('git-2.32.0-r0') }
    its('stdout') { should include ('openssh-8.6_p1-r2') }
    its('stdout') { should include ('tar-1.34-r0') }
    its('stdout') { should include ('gzip-1.10-r1') }
    its('stdout') { should include ('ca-certificates-20191127-r5') }
    its('stdout') { should include ('sudo-1.9.7_p1-r1') }
    its('stdout') { should include ('libintl-0.21-r0') }
  end
end

control 'git version' do
  impact 1.0
  title 'confirm git version installed'
  desc 'confirm version reported by git matches the desired version'
  describe command('git --version') do
    its('stdout') { should include ('2.32') }
  end
end

control 'openssh version' do
  impact 1.0
  title 'confirm openssh version installed'
  desc 'confirm version reported by openssh matches the desired version'
  describe command('ssh -V') do
    its('stderr') { should include ('8.6') }
  end
end

control 'tar version' do
  impact 1.0
  title 'confirm tar version installed'
  desc 'confirm version reported by tar matches the desired version'
  describe command('tar --version') do
    its('stdout') { should include ('1.34') }
  end
end

control 'gzip version' do
  impact 1.0
  title 'confirm gzip version installed'
  desc 'confirm version reported by gzip matches the desired version'
  describe command('gzip --version') do
    its('stdout') { should include ('1.10') }
  end
end

control 'ca-certificates installed' do
  impact 1.0
  title 'confirm ca-certificates installed'
  desc 'confirm certificate authorities have been installed for secure remote communication'
  describe command('ls /etc/ssl/certs/') do
    its('stdout') { should include ('ca-cert-DigiCert_Assured_ID_Root_CA.pem') }
  end
end

control 'sudo version' do
  impact 1.0
  title 'confirm sudo version installed'
  desc 'confirm version reported by sudo matches the desired version'
  describe command('sudo --version') do
    its('stdout') { should include ('1.9') }
  end
end

control 'localization' do
  impact 1.0
  title 'confirm locales for alpine'
  desc 'confirm localization installed'
  describe command('locale -a') do
    its('stdout') { should include ('C.UTF-8') }
    its('stdout') { should include ('fr_FR.UTF-8') }
    its('stdout') { should include ('de_CH.UTF-8') }
  end
  describe command('locale') do
    its('stdout') { should include ('LANG=C.UTF-8') }
    its('stdout') { should include ('LC_ALL=en_US.UTF-8') }
  end
end

describe user('circleci') do
  it { should exist }
  its('uid') { should eq 3434 }
  its('gid') { should eq 3434 }
  its('group') { should eq 'circleci' }
  its('home') { should eq '/home/circleci' }
end

describe directory('/home/circleci') do
  its('owner') { should eq 'circleci' }
  its('group') { should eq 'circleci' }
end

describe directory('/home/circleci/project') do
  its('owner') { should eq 'circleci' }
  its('group') { should eq 'circleci' }
end
